libdnf/rpm/solv/goal_private.cpp

```diff
@@ -166,49 +160,54 @@ static void same_name_subqueue(Pool * pool, Queue * in, Queue * out) {
         queue_push(out, queue_pop(in));
 }
 
-int sort_packages(const void * ap, const void * bp, void * s_cb) {
-    Id a = *(Id *)ap;
-    Id b = *(Id *)bp;
-    Pool * pool = ((struct InstallonliesSortCallback *)s_cb)->pool;
-    Id kernel = ((struct InstallonliesSortCallback *)s_cb)->running_kernel;
-    Solvable * sa = pool_id2solvable(pool, a);
-    Solvable * sb = pool_id2solvable(pool, b);
 
-    // if the names are different sort them differently, particular order does not matter as long as it's consistent.
-    int name_diff = sa->name - sb->name;
-    if (name_diff)
-        return name_diff;
+struct InstallonlyComparator {
+    InstallonlyComparator(Pool * pool, Id running_kernel) : pool(pool), running_kernel(running_kernel) {}
 
-    // same name, if one is/depends on the running kernel put it last
+    int operator()(const Id & a, const Id & b) {
+        Solvable * sa = pool_id2solvable(pool, a);
+        Solvable * sb = pool_id2solvable(pool, b);
 
-    // move available packages to end of the list
-    if (pool->installed != sa->repo)
-        return 1;
+        // if the names are different sort them differently, particular order does not matter as long as it's consistent.
+        int name_diff = sa->name - sb->name;
+        if (name_diff)
+            return name_diff;
 
-    if (pool->installed != sb->repo)
-        return -1;
+        // same name, if one is/depends on the running kernel put it last
 
-    if (kernel >= 0) {
-        if (a == kernel || can_depend_on(pool, sa, kernel))
+        // move available packages to end of the list
+        if (pool->installed != sa->repo)
             return 1;
-        if (b == kernel || can_depend_on(pool, sb, kernel))
-            return -1;
-        // if package has same evr as kernel try them to keep (kernel-devel packages)
-        Solvable * kernelSolvable = pool_id2solvable(pool, kernel);
-        if (sa->evr == kernelSolvable->evr) {
-            return 1;
-        }
-        if (sb->evr == kernelSolvable->evr) {
+
+        if (pool->installed != sb->repo)
             return -1;
+
+        if (running_kernel >= 0) {
+            if (a == running_kernel || can_depend_on(pool, sa, running_kernel))
+                return 1;
+            if (b == running_kernel || can_depend_on(pool, sb, running_kernel))
+                return -1;
+            // if package has same evr as kernel try them to keep (kernel-devel packages)
+            Solvable * kernelSolvable = pool_id2solvable(pool, running_kernel);
+            if (sa->evr == kernelSolvable->evr) {
+                return 1;
+            }
+            if (sb->evr == kernelSolvable->evr) {
+                return -1;
+            }
         }
+        return pool_evrcmp(pool, sa->evr, sb->evr, EVRCMP_COMPARE);
     }
-    return pool_evrcmp(pool, sa->evr, sb->evr, EVRCMP_COMPARE);
-}
+
+    Pool * pool = nullptr;
```

Comment:
Minor: `poll` is initialized in the constructor. No initialization is required here.